unidist/core/backends/mpi/core/controller/common.py

```diff
@@ -173,40 +209,88 @@ def _push_local_data(dest_rank, data_id):
         logger.debug("PUT LOCAL {} id to {} rank".format(data_id._id, dest_rank))
 
         mpi_state = communication.MPIState.get_instance()
         async_operations = AsyncOperations.get_instance()
         # Push the local master data to the target worker directly
+        if is_serialized:
+            operation_data = local_store.get_serialized_data(data_id)
+        else:
+            operation_data = {
+                "id": data_id,
+                "data": local_store.get(data_id),
+            }
+
         operation_type = common.Operation.PUT_DATA
-        if object_store.is_already_serialized(data_id):
-            serialized_data = object_store.get_serialized_data(data_id)
-            h_list, _ = communication.isend_complex_operation(
+        if is_blocking_op:
+            serialized_data = communication.send_complex_data(
                 mpi_state.comm,
-                operation_type,
-                serialized_data,
+                operation_data,
                 dest_rank,
-                is_serialized=True,
+                is_serialized,
             )
         else:
-            operation_data = {
-                "id": data_id,
-                "data": object_store.get(data_id),
-            }
             h_list, serialized_data = communication.isend_complex_operation(
                 mpi_state.comm,
                 operation_type,
                 operation_data,
                 dest_rank,
-                is_serialized=False,
+                is_serialized,
             )
-            object_store.cache_serialized_data(data_id, serialized_data)
-        async_operations.extend(h_list)
+            async_operations.extend(h_list)
+
+        if not is_serialized:
+            local_store.cache_serialized_data(data_id, serialized_data)
+
         #  Remember pushed id
-        object_store.cache_send_info(data_id, dest_rank)
+        local_store.cache_send_info(data_id, dest_rank)
+
+
+def _push_shared_data(dest_rank, data_id, is_blocking_op):
+    """
+    catiSend data loon associated with data ID to target rank using shared memory.
+
+    Parameters
+    ----------
+    dest_rank : int
+        Target rank.
+    data_id : unidist.core.backends.common.data_id.DataID
+        An ID to data.
+    is_blocking_op : bool
+        Whether the communication should be blocking or not.
+        If ``True``, the request should be processed immediatly
+        even for a worker since it can get into controller mode.
+    """
+    mpi_state = communication.MPIState.get_instance()
+    shared_store = SharedObjectStore.get_instance()
+    mpi_state = communication.MPIState.get_instance()
+    operation_type = common.Operation.PUT_SHARED_DATA
+    async_operations = AsyncOperations.get_instance()
+    info_package = shared_store.get_shared_info(data_id)
+    info_package["id"] = data_id
+    operation_data = info_package
+    if is_blocking_op:
+        communication.mpi_send_object(mpi_state.comm, operation_data, dest_rank)
```

Comment:
Why do we use `mpi_send_object` here? Shouldn't it be `send_simple_operation`?

Reply:
When worker received blocking `Get` request, it can send data using shared memory, but it does not need to send operation.